test/integration/test_tracing.py

```diff
@@ -0,0 +1,110 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+from __future__ import annotations
+
+import json
+import logging
+import time
+from typing import Any, Callable
+
+import httpx
+import jubilant
+
+
+def test_direct_connection(build_charm: Callable[[], str], juju: jubilant.Juju):
+    charm_path = build_charm()
+    juju.deploy(charm_path)
+    juju.integrate('tracing-tester', 'tempo')
+    status = juju.wait(jubilant.all_active)
+
+    spans = wait_spans(status.apps['tempo-worker'].address, test=lambda s: 'ops.main' in str(s))
+    assert 'ops.main' in [s['name'] for s in spans]
+
+    event_names = [e['name'] for e in get_events(spans)]
+    assert 'StartEvent' in event_names
+
+    checkpoint = time.time()
+    juju.run('tracing-tester/0', 'one', params={'arg': 'oneoneoneone'})
+    spans = wait_spans(
+        status.apps['tempo-worker'].address,
+        since=checkpoint,
+        test=lambda s: 'one_action' in str(s),
+    )
+
+    # A span generated by ops
+    assert 'one_action: TracingTesterCharm' in [s['name'] for s in spans]
+    # A span generated by the test charm
+    assert 'on action' in [s['name'] for s in spans]
+
+    action_span = next(s for s in spans if s['name'] == 'on action')
+    assert 'oneoneoneone' in json.dumps(action_span)
+
+
+def test_with_load_balancer(build_charm: Callable[[], str], juju: jubilant.Juju):
+    charm_path = build_charm()
+    print(charm_path)
+    pass
+
+
+def test_with_tls(build_charm: Callable[[], str], juju: jubilant.Juju):
+    pass
+
+
+def test_with_load_balancer_tls(build_charm: Callable[[], str], juju: jubilant.Juju):
+    pass
+
+
+def wait_spans(
+    address: str,
+    since: float = 0,
+    *,
+    test: Callable[[list[dict[str, Any]]], bool],
+    timeout: float = 60,
+):
+    deadline = time.time() + timeout
+    spans = []
+    while time.time() < deadline:
+        spans = get_spans(address, since)
+        if test(spans):
+            return spans
+        logging.info('waiting for spans, will retry')
+        time.sleep(1)
+    logging.warning('giving up on waiting for spans')
+    return spans
+
+
+def get_spans(address: str, since: float = 0):
+    with httpx.Client(base_url=f'http://{address}:3200/api/', timeout=5) as client:
+        r = client.get('search?tags=service.name=tracing-tester')
+        r.raise_for_status()
+        spans: list[dict[str, Any]] = []
+        for t in r.json()['traces']:
+            if float(t['startTimeUnixNano']) / 1e9 < since:
+                continue
+            r = client.get(f'traces/{t["traceID"]}')
```

Comment:
Is there any way to get more than one at once?

Reply:
Nope, there's no batch endpoint.

Note that there really isn't a "trace" object, rather, a trace is a series of batches (typically 1) with series of scope spans (typ 1) with series of spans (many) all of the latter sharing the same `.traceID` attribute.

In real world, a trace is distributed across microservices, while a span belongs to a single microservice. The spans are sent over in spare time, and there's no concept of "having received all spans that belong to trace", rather, "here are the spans I have received so far" and "is the root span finished", while child spans are allowed to outlive their parents and the root.